src/aws_encryption_sdk/materials_managers/default.py

```diff
@@ -101,6 +107,64 @@ def get_encryption_materials(self, request):
             signing_key=signing_key,
         )
 
+    def _get_encryption_materials_using_keyring(self, request):
+        """Creates encryption materials using underlying keyring.
+
+        :param request: encryption materials request
+        :type request: aws_encryption_sdk.materials_managers.EncryptionMaterialsRequest
+        :returns: encryption materials
+        :rtype: aws_encryption_sdk.materials_managers.EncryptionMaterials
+        :raises InvalidCryptographicMaterialsError: if keyring cannot complete encryption materials
+        :raises InvalidCryptographicMaterialsError:
+           if encryption materials received from keyring do not match request
+        """
+        algorithm = request.algorithm if request.algorithm is not None else self.algorithm
+        encryption_context = request.encryption_context.copy()
+
+        signing_key = self._generate_signing_key_and_update_encryption_context(algorithm, encryption_context)
+
+        expected_encryption_context = encryption_context.copy()
```

Comment:
Why not do the copy the same way you do below?
i.e. copy on pass?

Reply:
Dictionaries are mutable; I try to make sure that we don't feed any side-effects upstream or cause any side-effects downstream.

The copy on line 122 is taking a copy of the requested encryption context so that we can modify it without changing the value in the request.

```python
encryption_context = request.encryption_context.copy()
```

Then, on line 124, we modify it. Long-term I would like to refactor that method so that it returns the new encryption context rather than mutating the input, but mutation is how it has always been so I'm leaving that as-is for now to avoid feature-creep.

```python
signing_key = self._generate_signing_key_and_update_encryption_context(algorithm, encryption_context)
```

On line 126, we take a copy of the modified encryption context for later comparison.

```python
expected_encryption_context = encryption_context.copy()
```

Finally, on line 139, we check the encryption context in the final encryption materials against the copy we made on line 126, before we passed it to the keyring, to make sure that the keyring did not modify it.